3DSView/src/main/java/eu/livotov/labs/android/d3s/D3SView.java

```diff
@@ -231,51 +226,50 @@ public void setStackedMode(String stackedModePostbackUrl)
         this.stackedModePostbackUrl = stackedModePostbackUrl;
     }
 
-    private void completeAuthorization(String html)
+    private void completeAuthorizationIfPossible(String html)
     {
+        // If the postback has already been handled, stop now
         if (postbackHandled.get()) {
             return;
         }
 
+        // Try and find the MD and PaRes form elements in the supplied html
         String md = "";
         String pares = "";
 
-        Matcher localMatcher1 = mdFinder.matcher(html);
-        Matcher localMatcher2 = paresFinder.matcher(html);
-
-        if (localMatcher1.find())
-        {
-            md = localMatcher1.group(1);
-        }
-
-        if (localMatcher2.find())
-        {
-            pares = localMatcher2.group(1);
+        Matcher mdMatcher = mdFinder.matcher(html);
+        if (mdMatcher.find()) {
+            md = mdMatcher.group(1);
+        } else {
+            return; // Not Found
         }
-
-        if (!TextUtils.isEmpty(md))
-        {
-            Matcher valueMatcher = valuePattern.matcher(md);
-            if (valueMatcher.find())
-            {
-                md = valueMatcher.group(1);
-            }
+        
+        Matcher paresMatcher = paresFinder.matcher(html);
+        if (paresMatcher.find()) {
+            pares = paresMatcher.group(1);
+        } else {
+            return; // Not Found
         }
 
-        if (!TextUtils.isEmpty(pares))
-        {
-            Matcher valueMatcher = valuePattern.matcher(pares);
-            if (valueMatcher.find())
-            {
-                pares = valueMatcher.group(1);
-            }
+        // Now extract the values from the previously captured form elements
+        Matcher mdValueMatcher = valuePattern.matcher(md);
+        if (mdValueMatcher.find()) {
+        		md = mdValueMatcher.group(1);
+        } else {
+            return; // Not Found
         }
 
-        // If we didn't find both MD and PaRes, don't continue
-        if (TextUtils.isEmpty(md) || TextUtils.isEmpty(pares)) {
-            return;
+        Matcher paresValueMatcher = valuePattern.matcher(pares);
+        if (paresValueMatcher.find()){
+            pares = paresValueMatcher.group(1);
+        } else {
+            return; // Not Found
```

Comment:
Instead of returning after every else, check if `pares` and `md` are empty before attempting to invoke the `authorizationListener` that way we don't have to worry about four conditionals failing when we only need two values.

Reply:
This was meant as a performance improvement as we were running 4 regexes before that logic was hit previously. This method is called numerous times and the delay was noticeable. Doing the above helped. 

I'm open to improvement suggestions though (bearing the above in mind), as the above code is a bit rubbish really.